// i-chatbot/lib/chatbot/ChatBotUtil.js
'use strict';

exports.__esModule = true;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ChatBotUtil = function () {
  function ChatBotUtil() {
    _classCallCheck(this, ChatBotUtil);
  }

  ChatBotUtil.textMessage = function textMessage(text) {
    for (var _len = arguments.length, actions = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      actions[_key - 1] = arguments[_key];
    }

    return {
      type: 'text',
      isInbound: false,
      content: {
        text: text
      },
      actions: actions
    };
  };

  ChatBotUtil.userTextMessage = function userTextMessage(text) {
    return {
      type: 'text',
      isInbound: true,
      content: {
        text: text
      }
    };
  };

  ChatBotUtil.makeGetStartedButton = function makeGetStartedButton(text) {
    return {
      title: text,
      type: 'get-started'
    };
  };

  ChatBotUtil.makeReplyButton = function makeReplyButton(text, callback) {
    return {
      title: text,
      callback: callback,
      type: 'quick-reply'
    };
  };

  ChatBotUtil.makeTextInputField = function makeTextInputField(submit, placeholder, callback) {
    return {
      submit: submit,
      placeholder: placeholder,
      callback: callback,
      type: 'text-input'
    };
  };

  return ChatBotUtil;
}();

Array.prototype.any = function () {
  return this[Math.floor(Math.random() * this.length)];
};

exports.default = ChatBotUtil;
module.exports = exports['default'];
