// i-chatbot/es/chatbot/messages/Messages.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Scroll from 'react-scroll';
import { CSSTransitionGroup } from 'react-transition-group';

import Message from './message/Message';

var Element = Scroll.Element;
var Scroller = Scroll.scroller;

var Messages = function (_Component) {
  _inherits(Messages, _Component);

  function Messages() {
    _classCallCheck(this, Messages);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  Messages.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    if (this.props.messages.length === prevProps.messages.length + 1 || this.props.isTyping !== prevProps.isTyping) {
      Scroller.scrollTo('LastMessage', {
        duration: 500,
        smooth: true,
        containerId: 'Messages'
      });
    }
  };

  Messages.prototype.render = function render() {
    return React.createElement(
      CSSTransitionGroup,
      { component: 'ul',
        className: 'Messages',
        id: 'Messages',
        transitionName: 'Messages',
        transitionEnterTimeout: 500,
        transitionLeaveTimeout: 300 },
      this.props.messages.map(function (message, i) {
        return React.createElement(Message, _extends({}, message, { key: i }));
      }),
      React.createElement(Element, { name: 'LastMessage' })
    );
  };

  return Messages;
}(Component);

Messages.defaultProps = {
  messages: []
};

Messages.propTypes = process.env.NODE_ENV !== "production" ? {
  messages: PropTypes.arrayOf(PropTypes.object).isRequired
} : {};

export default Messages;
